File: src/main/java/com/meepoffaith/hextrapats/casting/actions/math/OpRandVec.java
package com.meepoffaith.hextrapats.casting.actions.math;

import at.petrak.hexcasting.api.casting.eval.CastingEnvironment;
import at.petrak.hexcasting.api.casting.iota.Iota;
import at.petrak.hexcasting.api.casting.iota.Vec3Iota;
import com.meepoffaith.hextrapats.casting.bases.ConstMediaActionBase;
import com.meepoffaith.hextrapats.casting.bases.HexIotaStack;
import net.minecraft.util.math.Vec3d;

import java.util.List;

public class OpRandVec extends ConstMediaActionBase{
    public int argc = 0;
    public long mediaCost = 0L;

    @Override
    public List<? extends Iota> execute(HexIotaStack stack, CastingEnvironment ctx){
        double z = -1 + 2 * ctx.getWorld().random.nextDouble();
        double r = Math.sqrt(1 - z * z);
        double t = ctx.getWorld().random.nextDouble() * 2 * Math.PI;
        return asActionResult(new Vec3Iota(new Vec3d(r * Math.cos(t), r * Math.sin(t), z)));
    }
}

File: src/main/kotlin/com/meepoffaith/hextrapats/casting/actions/vecmath/OpRandVec.kt
package com.meepoffaith.hextrapats.casting.actions.vecmath

import at.petrak.hexcasting.api.casting.asActionResult
import at.petrak.hexcasting.api.casting.castables.ConstMediaAction
import at.petrak.hexcasting.api.casting.eval.CastingEnvironment
import at.petrak.hexcasting.api.casting.iota.Iota
import net.minecraft.util.math.Vec3d

object OpRandVec : ConstMediaAction {
    override val argc = 0
    override fun execute(args: List<Iota>, env: CastingEnvironment): List<Iota> {
        val rand = env.world.random
        return Vec3d(rand.nextGaussian(), rand.nextGaussian(), rand.nextGaussian()).normalize().asActionResult
    }
}
